bin/write_viral_gff.py

```diff
@@ -21,136 +20,213 @@ def get_ena_contig_mapping(ena_contig_file):
+                checkv_type=line_l[2]
+                checkv_quality=line_l[7]
+                miuvig_quality=line_l[8]
+                value='checkv_provirus='+checkv_type+';checkv_quality='+checkv_quality+';miuvig_quality='+miuvig_quality
+                checkv_dict[contig_id]=value
 
     #   parse taxonomic lineage when available
     for tfile in taxonomy_files:
-        taxonomy_df = pd.read_csv(tfile, sep="\t")
-        for index, row in taxonomy_df.iterrows():
-            if row["contig_ID"] in virify_contigs:
-                contig_lineage = []
-                for lineage in [
-                    row["order"],
-                    row["family"],
-                    row["subfamily"],
-                    row["genus"],
-                ]:
-                    if "." in str(lineage) or pd.isna(lineage):
-                        contig_lineage.append("")
-                    else:
-                        contig_lineage.append(lineage)
-                joined_lineage = ";".join(contig_lineage)
-                #   set to unclassified if all levels are empty
-                if joined_lineage == ";;;":
-                    taxonomy_dict[row["contig_ID"]] = "unclassified"
+        with open(tfile,'r') as input_table:
+            next(input_table)
+            for line in input_table:
+                line_l=line.rstrip().split('\t')
+                Contig=line_l.pop(0)
+
+                if '|phage-circular' in Contig:
+                    Contig=Contig.replace('|phage-circular','')
+                elif '|prophage-' in Contig:
+                    desc='prophage'
                 else:
-                    taxonomy_dict[row["contig_ID"]] = joined_lineage
-
-    with open(output_filename, "w") as metadata:
-        metadata.write(headers)
-        for contig in virify_contigs:
-            #   change contig name if ena mapping required
-            if ena_mapping:
-                contig_name = f"{ena_mapping[contig]}-{contig}"
+                    Contig=Contig.replace('|','')
+
+                lineage=[]
+                for element in line_l:
+                    if len(element)>0:
+                        try:
+                            float(element)
+                        except:
+                            lineage.append(element)
+                if len(lineage)>0:
+                    taxo_string='%3B'.join(lineage)
+                else:
+                    taxo_string='unclassified'
+
+                taxonomy_dict[Contig]=taxo_string
+
+    # Saving the original contig length
+    for record in SeqIO.parse(assembly_file, "fasta"):
+        contig_id=str(record.id)
+        seq_len=len(str(record.seq))
+        contigs_len_dict[contig_id]=str(seq_len)
+
+    contig_protID=annot_dicts[0]
+    prot_data=annot_dicts[1]
+    contig_data=annot_dicts[2]
+    # Writing the gff file
+    with open(output_filename, "w") as gff:
+        gff.write("##gff-version 3\n")
+        contig_protID=annot_dicts[0]
```

Comment:
These 3 lines are repeated

Reply:
I've removed repeated lines and re-formatted as suggested in the next comment